File: app/src/java/java/com/example/ama/viewer/presentation/list/MainListFragment.java
package com.example.ama.viewer.presentation.list;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import com.example.ama.viewer.R;
import com.example.ama.viewer.data.repo.DataRepositoryImpl;
import com.example.ama.viewer.presentation.list.adapter.MainListAdapter;
import com.example.ama.viewer.presentation.list.mvp.MainListPresenterImpl;
import com.example.ama.viewer.presentation.list.mvp.base.MainPresenter;
import com.example.ama.viewer.presentation.list.mvp.base.MainView;
import com.hannesdorfmann.mosby3.mvp.viewstate.lce.LceViewState;
import com.hannesdorfmann.mosby3.mvp.viewstate.lce.MvpLceViewStateFragment;
import com.hannesdorfmann.mosby3.mvp.viewstate.lce.data.RetainingLceViewState;

import java.util.List;

import io.reactivex.android.schedulers.AndroidSchedulers;

public class MainListFragment
        extends MvpLceViewStateFragment<RecyclerView, List<String>, MainView, MainPresenter>
        implements MainView {

    private SwipeRefreshLayout srlLayout;
    private MainListAdapter adapter;

    public static MainListFragment newInstance() {
        return new MainListFragment();
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_main, container, false);
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        initUi(view);
        setListeners();
    }

    private void initUi(View view) {
        contentView.setLayoutManager(new LinearLayoutManager(getContext()));
        adapter = new MainListAdapter();
        contentView.setAdapter(adapter);
        srlLayout = view.findViewById(R.id.srl_layout);
    }

    private void setListeners() {
        srlLayout.setOnRefreshListener(() -> loadData(true));
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.main_menu, menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == R.id.mi_refresh && !viewState.isLoadingState()) {
            loadData(false);
            return true;
        }
        return false;
    }

    @Override
    public List<String> getData() {
        return adapter.getItems();
    }

    @NonNull
    @Override
    public MainPresenter createPresenter() {
        return new MainListPresenterImpl(
                new DataRepositoryImpl(),
                AndroidSchedulers.mainThread());
    }

    @NonNull
    @Override
    public LceViewState<List<String>, MainView> createViewState() {
        return new RetainingLceViewState<>();
    }

    @Override
    protected String getErrorMessage(Throwable e, boolean pullToRefresh) {
        if (pullToRefresh) srlLayout.setRefreshing(false);
        String message = e.getMessage();
        return message != null ? message : getResources().getString(R.string.unknown_error);
    }

    @Override
    public void loadData(boolean pullToRefresh) {
        getPresenter().loadData(pullToRefresh);
    }

    @Override
    public void showContent() {
        srlLayout.setRefreshing(false);
        super.showContent();
    }

    @Override
    public void setData(List<String> data) {
        adapter.setItems(data);
    }

    @Override
    protected void animateErrorViewIn() {
        contentView.setVisibility(View.GONE);
        errorView.setVisibility(View.VISIBLE);
        loadingView.setVisibility(View.GONE);
        srlLayout.setVisibility(View.VISIBLE);
    }

    @Override
    protected void animateContentViewIn() {
        contentView.setVisibility(View.VISIBLE);
        errorView.setVisibility(View.GONE);
        loadingView.setVisibility(View.GONE);
        srlLayout.setVisibility(View.VISIBLE);
    }

    @Override
    protected void animateLoadingViewIn() {
        super.animateLoadingViewIn();
        srlLayout.setVisibility(View.GONE);
    }

    @Override
    public void appendItemToList(String item) {
        adapter.appendItem(item);
    }
}

File: app/src/kotlin/java/com/example/ama/viewer/presentation/list/MainListFragment.kt
package com.example.ama.viewer.presentation.list

import android.os.Bundle
import android.support.v7.widget.CardView
import android.view.*
import android.widget.ImageView
import com.example.ama.viewer.R
import com.example.ama.viewer.ViewerApp
import com.example.ama.viewer.data.loader.ImageLoader
import com.example.ama.viewer.data.loader.PicassoImageLoader
import com.example.ama.viewer.data.model.GithubUser
import com.example.ama.viewer.data.repo.DataRepositoryImpl
import com.example.ama.viewer.presentation.list.mvp.MainListPresenterImpl
import com.example.ama.viewer.presentation.list.mvp.base.MainPresenter
import com.example.ama.viewer.presentation.list.mvp.base.MainView
import com.hannesdorfmann.mosby3.mvp.viewstate.lce.LceViewState
import com.hannesdorfmann.mosby3.mvp.viewstate.lce.MvpLceViewStateFragment
import com.hannesdorfmann.mosby3.mvp.viewstate.lce.data.RetainingLceViewState
import com.squareup.picasso.Picasso
import io.reactivex.android.schedulers.AndroidSchedulers
import kotlinx.android.synthetic.main.content_profile.*
import kotlinx.android.synthetic.main.fragment_profile.*

class MainListFragment : MvpLceViewStateFragment<CardView, GithubUser, MainView, MainPresenter>(), MainView {

    private var githubUser: GithubUser? = null
    private val imageLoader: ImageLoader<ImageView> = PicassoImageLoader(
            Picasso.Builder(ViewerApp.instance).build())

    companion object {
        fun newInstance() = MainListFragment()
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_profile, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        srl_layout.setOnRefreshListener { loadData(true) }
    }

    override fun onCreateOptionsMenu(menu: Menu?, inflater: MenuInflater?) {
        inflater?.inflate(R.menu.main_menu, menu)
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        if (item?.itemId == R.id.mi_refresh && !viewState.isLoadingState) {
            loadData(false)
            return true
        }
        return false
    }

    override fun loadData(pullToRefresh: Boolean) {
        presenter.loadData(pullToRefresh)
    }

    override fun createPresenter(): MainPresenter =
            MainListPresenterImpl(
                    DataRepositoryImpl(ViewerApp.instance.githubApi),
                    AndroidSchedulers.mainThread())

    override fun createViewState(): LceViewState<GithubUser, MainView> =
            RetainingLceViewState<GithubUser, MainView>()

    override fun setData(data: GithubUser) {
        this.githubUser = data
        if (data.avatar != null) {
            imageLoader.loadImage(data.avatar, iv_avatar)
        }
        tv_login.text = data.login
        tv_content_name.text = data.name
        tv_content_blog.text = data.blog
        tv_content_company.text = data.company
        tv_content_email.text = data.email
        tv_content_location.text = data.location
        tv_content_bio.text = data.bio
    }

    override fun getData() = githubUser

    override fun getErrorMessage(e: Throwable?, pullToRefresh: Boolean): String {
        if (pullToRefresh) srl_layout.isRefreshing = false
        return e?.message ?: resources.getString(R.string.unknown_error)
    }

    override fun showContent() {
        srl_layout.isRefreshing = false
        super.showContent()
    }

    override fun animateErrorViewIn() {
        contentView.visibility = View.GONE
        errorView.visibility = View.VISIBLE
        loadingView.visibility = View.GONE
        srl_layout.visibility = View.VISIBLE
    }

    override fun animateContentViewIn() {
        contentView.visibility = View.VISIBLE
        errorView.visibility = View.GONE
        loadingView.visibility = View.GONE
        srl_layout.visibility = View.VISIBLE
    }

    override fun animateLoadingViewIn() {
        super.animateLoadingViewIn()
        srl_layout.visibility = View.GONE
    }
}